crates/network/src/rpc_router.rs

```diff
@@ -0,0 +1,412 @@
+    pub fn generate_token(&self) -> Option<RpcToken<T>>
+    where
+        T::CorrelationId: Default,
+    {
+        let correlation_id = T::CorrelationId::default();
+        self.new_token(correlation_id)
+    }
+
+    /// Handles the next message. This will **return** the message if no correlated request is
+    /// in-flight. otherwise, it's handled by the corresponding token receiver.
+    pub async fn handle_next_or_get(&mut self) -> Option<MessageEnvelope<T>> {
+        tokio::select! {
+            Some(message) = self.incoming_messages.next() => {
+                self.flight_tracker.handle_message(message)
+            },
+            _ = cancellation_watcher() => { None },
+            else => { None } ,
+        }
+    }
+}
+
+struct RpcTokenSender<T> {
+    sender: oneshot::Sender<MessageEnvelope<T>>,
+}
+
+pub struct RpcToken<T>
+where
+    T: RpcMessage,
+{
+    correlation_id: T::CorrelationId,
+    router: Weak<Inner<T>>,
+    // This is Option to get around Rust's borrow checker rules when a type implements the Drop
+    // trait. Without this, we cannot move receiver out.
+    receiver: Option<oneshot::Receiver<MessageEnvelope<T>>>,
+}
+
+impl<T> RpcToken<T>
+where
+    T: RpcMessage,
+{
+    pub fn correlation_id(&self) -> T::CorrelationId {
+        self.correlation_id.clone()
+    }
+
+    /// Awaits the response to come for the associated request. Cancellation safe.
+    pub async fn recv(mut self) -> Result<MessageEnvelope<T>, ShutdownError> {
+        let receiver = std::mem::take(&mut self.receiver);
+        let res = match receiver {
+            Some(receiver) => {
+                tokio::select! {
+                    _ = cancellation_watcher() => {
+                        return Err(ShutdownError);
+                    },
+                    res = receiver => {
+                        res.map_err(|_| ShutdownError)
+                    }
+                }
+            }
+            // Should never happen unless token was created with None which shouldn't be possible
+            None => Err(ShutdownError),
+        };
+        // If we have received something, we don't need to run drop() since the flight tracker has
+        // already removed the sender token.
+        std::mem::forget(self);
+        res
+    }
+}
+
+impl<T> Drop for RpcToken<T>
+where
+    T: RpcMessage,
+{
+    fn drop(&mut self) {
+        // if the router is gone, we can't do anything.
+        let Some(router) = self.router.upgrade() else {
+            return;
+        };
+        let _ = router.in_flight.remove(&self.correlation_id);
+    }
+}
+
+impl<T> MessageHandler for ResponseTracker<T>
+where
+    T: RpcMessage + WireDecode + Targeted,
+{
+    type MessageType = T;
+
+    fn on_message(
+        &self,
+        msg: restate_node_protocol::MessageEnvelope<Self::MessageType>,
+    ) -> impl std::future::Future<Output = ()> + Send {
+        self.handle_message(msg);
```

Comment:
I'm not sure about that one. I suspect that it can be a common place where the tracker token is dropped on timeouts and it would be noisy (and potentially not very useful) log message in this case.